From: Simon McVittie <smcv@debian.org>
Date: Wed, 11 Aug 2021 14:23:12 +0100
Subject: demo: Add --timed-quit argument

This allows the demo to be run as a non-interactive smoke test.

Signed-off-by: Simon McVittie <smcv@debian.org>
Forwarded: https://gitlab.gnome.org/jadahl/libdecor/-/merge_requests/79
Applied-upstream: 0.1.1, commit:771b55004b6fff0ce121811a55c2ef08abea6e3e
---
 demo/demo.c | 21 ++++++++++++++++++++-
 1 file changed, 20 insertions(+), 1 deletion(-)

diff --git a/demo/demo.c b/demo/demo.c
index 3cfdc4b..5332e82 100644
--- a/demo/demo.c
+++ b/demo/demo.c
@@ -1239,10 +1239,19 @@ int
 main(int argc,
      char **argv)
 {
+	struct timeval start;
+	struct timeval now;
 	struct wl_display *wl_display;
 	struct wl_registry *wl_registry;
 	struct libdecor *context;
 	struct output *output;
+	int timeout = -1;
+	bool timed_quit = false;
+
+	if (argc > 1 && strcmp(argv[1], "--timed-quit") == 0) {
+		timed_quit = true;
+		timeout = 500; /* ms */
+	}
 
 	/* write all output to stdout immediately */
 	setbuf(stdout, NULL);
@@ -1284,7 +1293,17 @@ main(int argc,
 	libdecor_frame_set_title(window->frame, "libdecor demo");
 	libdecor_frame_map(window->frame);
 
-	while (libdecor_dispatch(context, -1) >= 0);
+	gettimeofday(&start, NULL);
+
+	while (libdecor_dispatch(context, timeout) >= 0) {
+		if (timed_quit) {
+			gettimeofday(&now, NULL);
+			if (now.tv_sec >= start.tv_sec + 2) {
+				fprintf(stderr, "Exiting due to --timed-quit\n");
+				libdecor_frame_close(window->frame);
+			}
+		}
+	}
 
 	free_outputs();
 
